lwt/xen_api_lwt_unix.ml

```diff
@@ -62,29 +62,14 @@ module Lwt_unix_IO = struct
     Ssl.create_context Ssl.TLSv1_2 Ssl.Client_context
 
   let open_connection uri =
-    let domain_addr_t = match Uri.host uri with
-      | Some host ->
-        Lwt.catch
-          (fun () ->
-             Lwt_unix.gethostbyname host >>= fun host_entry ->
-             return (host_entry.Lwt_unix.h_addrtype, host_entry.Lwt_unix.h_addr_list.(0)))
-          (fun _ -> fail (Failed_to_resolve_hostname host))
-      | None -> fail (Failed_to_resolve_hostname "") in
-    (match Uri.scheme uri with
-     | Some "http" -> return false
-     | Some "https" -> return true
-     | Some "file" -> return false
-     | Some x -> fail (Unsupported_scheme x)
-     | None -> fail (Unsupported_scheme "")) >>= fun ssl ->
-    let port = match Uri.port uri with
-      | Some x -> x
-      | None -> if ssl then 443 else 80 in
+    Lwt.catch (fun () -> return (Util.is_ssl uri)) (fun _ -> match Uri.scheme uri with
+      | Some x -> fail (Unsupported_scheme x)
+      | None -> fail (Unsupported_scheme "")) >>= fun ssl ->
     (match Uri.scheme uri with
      | Some "file" ->
        return (Unix.PF_UNIX, Unix.ADDR_UNIX (Uri.path uri))
      | Some "http" | Some "https" ->
-       domain_addr_t >>= fun (domain, addr) ->
-       return (domain, Unix.ADDR_INET(addr, port))
+      Util.sockaddr_of_uri uri >|= fun sockaddr -> ((Unix.domain_of_sockaddr sockaddr), sockaddr)
      | _ -> assert false) >>= fun (domain, sockaddr) ->
```

Comment:
While simplified, the code is still tough to understand. I think a single match could work, if the change above is adopted:
```suggestion
    match Uri.scheme uri with
     | Some "file" ->
       return (Unix.PF_UNIX, Unix.ADDR_UNIX (Uri.path uri))
     | Some "http" | Some "https" ->
       Util.sockaddr_of_uri uri >|= fun (sockaddr, ssl) -> ((Unix.domain_of_sockaddr sockaddr), sockaddr)
     | Some x -> fail (Unsupported_scheme x)
     | None -> fail (Unsupported_scheme "") >>= fun (domain, sockaddr) ->
```
Note that sockaddr_of_uri won't fail as the uri has already been matched for http(s)

Reply:
Done!